pkg/virt-launcher/virtwrap/network/network.go

```diff
@@ -0,0 +1,344 @@
+	AddrDel(link netlink.Link, addr *netlink.Addr) error
+	AddrAdd(link netlink.Link, addr *netlink.Addr) error
+	LinkSetDown(link netlink.Link) error
+	LinkSetUp(link netlink.Link) error
+	LinkAdd(link netlink.Link) error
+	ParseAddr(s string) (*netlink.Addr, error)
+	ChangeMacAddr(iface string) (net.HardwareAddr, error)
+	GetMacDetails(iface string) (net.HardwareAddr, error)
+	StartDHCP(vm *v1.VirtualMachine, nic *VIF, serverAddr *netlink.Addr)
+}
+
+type NetworkUtilsHandler struct{}
+
+var Handler NetworkHandler
+
+func (h *NetworkUtilsHandler) LinkByName(name string) (netlink.Link, error) {
+	return netlink.LinkByName(name)
+}
+func (h *NetworkUtilsHandler) AddrList(link netlink.Link, family int) ([]netlink.Addr, error) {
+	return netlink.AddrList(link, family)
+}
+func (h *NetworkUtilsHandler) RouteList(link netlink.Link, family int) ([]netlink.Route, error) {
+	return netlink.RouteList(link, family)
+}
+func (h *NetworkUtilsHandler) AddrDel(link netlink.Link, addr *netlink.Addr) error {
+	return netlink.AddrDel(link, addr)
+}
+func (h *NetworkUtilsHandler) LinkSetDown(link netlink.Link) error {
+	return netlink.LinkSetDown(link)
+}
+func (h *NetworkUtilsHandler) LinkSetUp(link netlink.Link) error {
+	return netlink.LinkSetUp(link)
+}
+func (h *NetworkUtilsHandler) LinkAdd(link netlink.Link) error {
+	return netlink.LinkAdd(link)
+}
+func (h *NetworkUtilsHandler) ParseAddr(s string) (*netlink.Addr, error) {
+	return netlink.ParseAddr(s)
+}
+func (h *NetworkUtilsHandler) AddrAdd(link netlink.Link, addr *netlink.Addr) error {
+	return netlink.AddrAdd(link, addr)
+}
+
+// GetMacDetails from an interface
+func (h *NetworkUtilsHandler) GetMacDetails(iface string) (net.HardwareAddr, error) {
+	currentMac, err := lmf.GetCurrentMac(iface)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to get mac information for interface: %s", iface)
+		return nil, err
+	}
+	return currentMac, nil
+}
+
+// ChangeMacAddr changes the MAC address for a agiven interface
+func (h *NetworkUtilsHandler) ChangeMacAddr(iface string) (net.HardwareAddr, error) {
+	var mac net.HardwareAddr
+
+	currentMac, err := Handler.GetMacDetails(iface)
+	if err != nil {
+		return nil, err
+	}
+
+	changed, err := lmf.SpoofMacRandom(iface, false)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to spoof MAC for iface: %s", iface)
+		return nil, err
+	}
+
+	if changed {
+		mac, err = Handler.GetMacDetails(iface)
+		if err != nil {
+			return nil, err
+		}
+		log.Log.Reason(err).Errorf("Updated Mac for iface: %s - %s", iface, mac)
+	}
+	return currentMac, nil
+}
+
+func (h *NetworkUtilsHandler) StartDHCP(vm *v1.VirtualMachine, nic *VIF, serverAddr *netlink.Addr) {
+	// panic in case the DHCP server failed during the vm creation
+	// but ignore dhcp errors when the vm is destroyed or shutting down
+	if err := DHCPServer(
+		nic.MAC,
+		nic.IP.IP,
+		nic.IP.Mask,
+		macVlanIfaceName,
+		serverAddr.IP,
+		nic.Gateway,
+		net.ParseIP(guestDNS),
+	); err != nil && !vm.IsFinal() {
```

Comment:
I don't think this is going to do what you're intending it to do. 

The VM object passed in to this function won't reflect the actual state of the VM when the DHCP server exits.

Reply:
Why is the vm even needed here? We need to make sure that we don't even run that code again.